### PR TITLE
master_june24 completion: nb_warp_used (and a SUBset of gpucpp_june24) into gpucpp

### PR DESCRIPTION
**Pull request by valassi**
Hi @oliviermattelaer this is a WIP MR into gpucpp of my nb_warp_used and of everything else I am using against master_june24 in my (https://github.com/madgraph5/madgraph4gpu/pull/882).

The nb_warp_used fixes are described here https://github.com/madgraph5/madgraph4gpu/issues/885#issuecomment-2233827188

NB: as discussed in https://github.com/madgraph5/madgraph4gpu/issues/886, gpucpp_june24, with or without my changes, does NOT work for me with master_june24 (my modified version, but I am pretty sure that the problem was pre-existing). This is why here I am NOT merging gpucpp_june24, I am only merging a SUBSET of that (what was originally used in master_june24, plus my changes, see https://github.com/valassi/mg5amcnlo/commits/valassi_gpucpp_june24/

I keep this in WIP for the moment. I would only merge this when we agree to merge my https://github.com/madgraph5/madgraph4gpu/pull/882 into master (as master is meant to use the head of gpucpp... I do not want to 'pollute' gpucpp before!).

I also filed PR #120 against gpucpp_june24, but as mentioned in https://github.com/madgraph5/madgraph4gpu/issues/886, **I am unable to use gpucpp_june24 and this is NOT what was used in master_june24** anyway.

Thanks Andrea

### CODE REVIEW

**Comment by valassi** [comment]
Hi @oliviermattelaer this is a preliminary step for merging the master_june24 work of https://github.com/madgraph5/madgraph4gpu/pull/882

Can you please review and approve?

Note, this is related to https://github.com/madgraph5/madgraph4gpu/issues/886: in practice, what is here my valassi_gpucpp_june24 gets merged into gpucpp, while we forget about gpucpp_june24.

Thanks!
Andrea

**Comment by valassi** [comment]
PS @oliviermattelaer PLEASE DO NOT SQUASH in case you merge this, thanks :-)

**Comment by valassi** [comment]
Hi @oliviermattelaer I just merged th elatest gpucpp (including your gpucpp_fixci #132) into my valassi_gpucpp_june24.

I will update also my https://github.com/madgraph5/madgraph4gpu/pull/882 in madgraph4gpu to use this one.

(Then I will probably comment on your https://github.com/madgraph5/madgraph4gpu/pull/981 that 981 becomes irrelevant)

**Comment by valassi** [comment]
Hi @oliviermattelaer this now includes everything we need for june24. Can you have a look and approve? Thanks
Andrea
PS Again please do not squash this...

**Comment by oliviermattelaer** [comment]
Looks like they are conflict to fix here.
Additionally don't you want to use nb_warp_used? in the file dsample.f 
Or do you prefere anoter PR for that?

Olivier

File: madgraph/iolibs/export_v4.py [hunk 1]
```diff
@@ -1738,7 +1738,25 @@ def get_pdf_lines(self, matrix_element, ninitial, subproc_group = False, vector=
         if vector:
             pdf_definition_lines_vec = ""
             pdf_data_lines_vec = ""
-            pdf_lines = " DO iVEC=1,VECSIZE_USED\n"
+            pdf_lines = """ NB_WARP_USED = VECSIZE_USED / WARP_SIZE
```
Comment (valassi) [inline madgraph/iolibs/export_v4.py:1741]: Hi @oliviermattelaer about your question: I am doing nb_warp_used in various places. This is one example here.

But you are right that there may be more. (See https://github.com/madgraph5/madgraph4gpu/issues/983)

Ok I will check dsample.f more in detail...

**Comment by valassi** [comment]
Hi Olivier thanks

> Looks like they are conflict to fix here.

Hm? No I do not see conflicts, which conflicts are you talking about?

> Additionally don't you want to use nb_warp_used? in the file dsample.f
> Or do you prefere anoter PR for that?

No I wanted to fix this here.

I am fixing it in various places. 
See https://github.com/mg5amcnlo/mg5amcnlo/pull/121/files#r1740773122

I thought that this was enough because tests were passing.
But you are right that I may be missing some and that these go undetected.
Ok then lets put this back to WIP :-(

Andrea

**Comment by oliviermattelaer** [comment]
Screenshot 2024-09-02 at 13 56 21

Ok looks like this is another subtelty of github, that I had to understand... But this is only because the button was on "rebase" if I move it to "merge" then github does not complain about conflict anymore ...
Sorry for the confusion

**Comment by valassi** [comment]
> Additionally don't you want to use nb_warp_used? in the file dsample.f

Hi @oliviermattelaer again. I had a look at this, see the comments in https://github.com/madgraph5/madgraph4gpu/issues/983. In my opinion there is nothing else to be done in dsample.f, but I may be missing something. Can you have a look yourself and comment in 983 please?

Then if we agree that there is nothing else to do , I think that we can start merging this.

Thanks Andrea

**Comment by valassi** [comment]
> Hi @oliviermattelaer again. I had a look at this, see the comments in [madgraph5/madgraph4gpu#983](https://github.com/madgraph5/madgraph4gpu/issues/983). In my opinion there is nothing else to be done in dsample.f, but I may be missing something. Can you have a look yourself and comment in 983 please?

Hi @oliviermattelaer thanks for having a look! You confirmed here https://github.com/madgraph5/madgraph4gpu/issues/983#issuecomment-2325232569 that there is nothing to do for 983, so I consider that this is good to go.

So, I AM MERGING THIS! Starting the full june24 merge into master+gpucpp...

**Pull request merged**